### PR TITLE
setting up initial foundations for siteSR

### LINKED ISSUE #2: Pivot to using the NHDPlusHR MapServer query instead of medium resolution via {nhdplusTools}
See the example here: https://github.com/rossyndicate/ROSS_mini_tools/pull/25

### LINKED ISSUE #3: Add 'HUCEightDigitCode_flag' for instances where we have gapfilled
At this moment, this is just filled in without any indication that we've done this.

### PR DESCRIPTION
**Pull request by steeleb**
Hey MB and KW!

Tagging you both in this PR to establish some... "wet clay" ...code. Trying to set up foundational architecture for our siteSR product and this is the initial bit of that does the following:

In _targets.R:

1. retrieves the site info from Drive (only the chla exists for you all, but I've attached the .csv of the preliminary output from SDD here: [sdd_drive_ids.csv](https://github.com/user-attachments/files/16700620/sdd_drive_ids.csv) - if you plop this in the AquaMatch_harmonize_WQP/3_harmonize/out folder, the code should run without issue for you these are drive_ids associated with my googledrive, not the ROSSyndicate)

In 4_compile_sites.R:

2. collates sites (target: p4_distinct_sites) - I've added some flexibility here for folks who don't want to do the remote sensing query for sites not in their parameter of interest. Parameters are defined in the config.yml file.

3. converts all sites to a common CRS (target: p4_harmonized_sites) using the code @mbrousil wrote for determining timezones. MB, note I added error handling in here for when CRS exist in the location list, but don't exist in our translate table. Feel free to copy this over to the harmonize pipeline.

4. target p4_add_HUC8 fills in the HUCEightDigitCode column as needed (<1% of sites are missing this, but attributing the point to a waterbody or flowline is easier when we have this information)

5. target p4_add_NHD_waterbody_info attributes a NHDPlusHR permanent identifier to each point, if it intersects with a NHD waterbody polygon

6. the p4_add_NHD_waterbody_info target also uses an incredilby simple method to add some variable of confidence in our assignment of that comid. For this, I just add a 100m buffer to the point and see how many polygons the buffer intersects, tallying them in the column n_wbd_100m

7. Number 5&6 are repeated for flowline attribution as p4_add_NHD_flowline. In this case, we're relating a point to a line, so sf::st_closest() is used, but we use the same method of uncertainty (100m buffer and sf::st_join()).

8. Aggregate all of the site info in a single dataframe: p4_sites_with_NHD_attribution. 

I think that the p4_add_NHD_waterbody_info and p4_add_NHD_flowline_info might be a starting place for the EPA_TADA request about adding some measure of uncertainty around attributing NHD information to WQP points... I don't remember what TADA issue this is related to, but remember it being an ask of us.

Known issues:

- This is computationally expensive. It's about 1h to add the waterbody information to the sites and more than 1h to add flowline attribution. This isn't the end of the world, but a pain. Moving to processing by HUC 4 speeds up the process by half (#4), but if you see any spots to speed this up, I welcome suggestions to speed this up further (but maybe it is what it is?)
- I've used the NHDPlusHR using the MapServer for this workflow (https://github.com/rossyndicate/ROSS_mini_tools/pull/25). There are still some HUCs that don't get pulled for one reason or another for waterbody and flowline attribution. These are not always the same between the two feature types (at this impasse, I don't know why, but will eventually figure it out). For now, this is noted in issue #17 

@mbrousil could you review for general consistency with AquaMatch and eagle eye this? @kathryn-willi I'm tagging you for your expertise in NHD land, so a focus on the add_NHD_waterbody_to_sites/add_NHD_flowline_to_sites would be great.

Note, I've only run this using the admin_update config. I also acknowledge this is a pretty dense PR and y'all have some deadlines you're up against - is a review by Friday possible?

Progress on #6, #7, Closes #2 (functions all use NHDPlusHR), Closes #3 (which was left over from the closed PR from last week)

### CODE REVIEW

File: 4_compile_sites/src/add_NHD_flowline_to_sites.R [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      # transform the points to the same crs as the flowlines
+      sf_subset <- st_transform(sf_subset, crs = st_crs(flowlines_subset))
+      # get the rowid of the closest flowline
+      sf_subset$closest_flow_rowid <- st_nearest_feature(sf_subset, flowlines_subset)
+      df_subset_with_flow <- sf_subset %>% 
+        split(f = .$rowid) %>% 
+        map(.x = .,
+            .f = ~{
+              # calculate distance
+              .x$dist_to_flow <- st_distance(.x, flowlines_subset[.x$closest_flow_rowid, ])
```
Comment (kathryn-willi) [inline 4_compile_sites/src/add_NHD_flowline_to_sites.R:51]: Is it possible that st_nearest() could grab a flowline very far away with this method, if there are no flowlines nearby? What about using a 100m buffer first to grab flowlines nearby, then performing st_nearest on that subset? In a perfect world WQP obs are occurring near a flowline but there are definitely areas of the NHD that are spotty/surface water features are poorly mapped.

File: 4_compile_sites/src/add_NHD_flowline_to_sites.R [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+              .x
+            }) %>% 
+        bind_rows() %>% 
+        select(-c(closest_flow_rowid)) %>% 
+        st_drop_geometry()
+      
+      # add a 100m buffer and see how many points intersect multiple flowlines
+      # this is a stand in for how 'confident' we are in assigning a given point to a
+      # flowline
+      df_buffer_flow <- sf_subset %>%
```
Comment (kathryn-willi) [inline 4_compile_sites/src/add_NHD_flowline_to_sites.R:63]: I like this quick "confidence" approach!

File: 4_compile_sites/src/harmonize_crs.R [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  # Add EPSG codes
+  site_w_epsg <- sites %>%
+    left_join(x = .,
+              y = epsg_codes,
+              by = "HorizontalCoordinateReferenceSystemDatumName") 
+  
+  # Check to see if there are any sites that have CRS not in the tribble above,
+  # if so, stop and message!
+  if (nrow(filter(site_w_epsg, is.na(epsg))) > 0) {
+    stop("There is at least one CRS datum in the site list that is not included
```
Comment (kathryn-willi) [inline 4_compile_sites/src/harmonize_crs.R:63]: Or... could they just default to WGS84?

**Review by kathryn-willi — APPROVED** [review]
NOTE: I did NOT attempt to run this pipeline, so my comments/suggestions are untested by me.

Is there a way to query across HUC2s? Just thinking if HUC4 is faster than HUC8, might be faster to pull HUC2? Though those areas are huge, so if there aren't that many obs in a single HUC2 might be slower than what you get out of it...

https://www.usgs.gov/media/images/nhdplus-hr-status-map-0 Just looking at HUC4s without features... there is definitely _some_ overlap between HUC4s that don't have NHDPlus HR yet (possible never with 3DHP up and running now?) and those listed... For those that don't have HiRes that aren't in Alaska, can it use NHDPlus V2 as a backup? And then for Alaska... I really have no idea what can be used... nhdplusTools now has a way of pulling in 3DHP for user-specified points: https://doi-usgs.github.io/nhdplusTools/articles/get_3dhp_data.html. Dave Blodgett told me it is even available for Alaska now (https://github.com/DOI-USGS/nhdplusTools/issues/353)?

File: 4_compile_sites/src/add_HUC8_to_sites.R [hunk 4]
```diff
@@ -0,0 +1,50 @@
+#' @title Add HUC8 information to sites if not populated
+#' 
+#' @description
+#' This funciuon uses the location of a site to determine the HUC8 it
```
Comment (mbrousil) [inline 4_compile_sites/src/add_HUC8_to_sites.R:4]:
```suggestion
#' This function uses the location of a site to determine the HUC8 it
```

File: _targets.R [hunk 5]
```diff
@@ -6,17 +6,15 @@ library(tarchetypes)
 
 # Set target options:
 tar_option_set(
-  packages = "tidyverse",
+  packages = c("tidyverse","sf"),
```
Comment (mbrousil) [inline _targets.R:9]: nice lol

**Review by mbrousil — APPROVED** [review]
Successfully ran the pipeline! Took ~4.5 hours.

File: 4_compile_sites.R [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  # 0 = HUC8 reported in WQP site information
+  # 1 = HUC8 determined from NHDPlusHR
+  # 2 = HUC8 unable to be determined for site location
+  tar_target(
+    name = p4_add_HUC8,
+    command = {
+      need_HUC8 <- p4_harmonized_sites %>%
+        filter(is.na(HUCEightDigitCode)) %>% 
+        mutate(flag_HUC8 = 1)
+      assigned_HUC8 <- add_HUC8_to_sites(sites_without_HUC = need_HUC8) %>%
```
[the remark this answers is not in the document]
Comment (steeleb) [reply 4_compile_sites.R:59]: I didn't! I'm sure that nhdplusTools::get_huc() is faster, but it references NHDPlusV2. I had assumed that the HUC ids were different between NHDPlusV2 and NHDPlusHR. But maybe that's not the case?

File: 4_compile_sites/src/add_NHD_flowline_to_sites.R (hunk 1 above)
Comment (steeleb) [reply 4_compile_sites/src/add_NHD_flowline_to_sites.R:51]: Yes, and that's why we include the distance, too! Do you think between this and the confidence approach, that I need to address this now, or can this be part of later documentation?

File: 4_compile_sites/src/harmonize_crs.R (hunk 3 above)
Comment (steeleb) [reply 4_compile_sites/src/harmonize_crs.R:63]: Similar to the suggestion with TADA - warn, but assign WGS84? I think for us, I'd really like to know when these instances happen and fix it, but for downstream users, it should default to WGS84. What do you think about pulling in the config setting here with an if/else for error handling?

File: config.yml [hunk 7]
```diff
@@ -4,36 +4,31 @@
...
   google_email: ""
-  # The local directory where the first {targets} pipeline is located (i.e.,
+  # The local directory where the harmonize {targets} pipeline is located (i.e.,
   # the pipeline that runs the download step)
-  download_repo_directory: "../AquaMatch_harmonize_WQP/"
-  # Use a stable version of the harmonized parameter datasets?
-  chl_use_stable: FALSE
-  doc_use_stable: FALSE
+  harmonize_repo_directory: "../AquaMatch_harmonize_WQP/"
+  parameters: "chla,sdd" # any combination of harmonized parameters in lowercase separated by ","
```
[the remark this answers is not in the document]
Comment (steeleb) [reply config.yml:10]: Absolutely agree.

**Comment by steeleb** [comment]
Thanks @kathryn-willi!

> Is there a way to query across HUC2s? Just thinking if HUC4 is faster than HUC8, might be faster to pull HUC2? Though those areas are huge, so if there aren't that many obs in a single HUC2 might be slower than what you get out of it...

I think this is the balance - I do suspect that HUC2s are too large and would query too many features (I think the max is 2k per MapServer pull). This leads me to your next comment!

> https://www.usgs.gov/media/images/nhdplus-hr-status-map-0 Just looking at HUC4s without features... there is definitely some overlap between HUC4s that don't have NHDPlus HR yet (possibly never with 3DHP up and running now?) and those listed... For those that don't have HiRes that aren't in Alaska, can it use NHDPlus V2 as a backup? And then for Alaska... I really have no idea what can be used... nhdplusTools now has a way of pulling in 3DHP for user-specified points: https://doi-usgs.github.io/nhdplusTools/articles/get_3dhp_data.html. Dave Blodgett told me it is even available for Alaska now (https://github.com/DOI-USGS/nhdplusTools/issues/353)?

I honestly haven't even looked at them (playing MVP here), I know some are because they are in an area where HR doesn't exist, I suspect that others are due to the 2k limit on MapServer. The 3dhp is pretty spotty (and [AK is still only partial](https://www.usgs.gov/media/images/3d-hydrography-program-fy24-partnerships)) so I think I'd rather pull from NHD best res [like here](https://github.com/steeleb/lakeSR/blob/main/a_Calculate_Centers/src/calculate_bestres_centers.R) for AK. For things due to the MapServer query limit, I'll probably end up splitting the HUC4 AOI by HUC8 to reduce the query size. If for some reason, this still doesn't work I'll definitely be defaulting to NHDPlusV2.

LMK if this all seems amenable!

File: 4_compile_sites.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (kathryn-willi) [reply 4_compile_sites.R:59]: HUCs technically come from the Watershed Boundary Dataset, which I believe (especially at the HUC4 resolution) aren't different across NHD versions...

File: 4_compile_sites/src/add_NHD_flowline_to_sites.R (hunk 1 above)
Comment (kathryn-willi) [reply 4_compile_sites/src/add_NHD_flowline_to_sites.R:51]: Nah! I think this is good especially for now. Something in the final documentation would def be useful tho!

File: 4_compile_sites/src/harmonize_crs.R (hunk 3 above)
Comment (kathryn-willi) [reply 4_compile_sites/src/harmonize_crs.R:63]: I like that idea! But it isn't super important if you don't feel like incorporating.

**Comment by kathryn-willi** [comment]
Good to go!

File: 4_compile_sites.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (steeleb) [reply 4_compile_sites.R:59]: This is awesome news. I'll do some testing to see if it's any faster! (And maybe a smidge of detective work to see if there are any iterations of WBD that correlate with the NHD versions.)

File: 4_compile_sites/src/harmonize_crs.R (hunk 3 above)
Comment (steeleb) [reply 4_compile_sites/src/harmonize_crs.R:63]: I'll make an issue for it and label it as low priority! #20

**Pull request merged**